core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,300 @@
+      if (!resources.release(seekableInputStream.getResourceId())) {
+        LOG.debug("Close the expired input stream resource of {}",
+            seekableInputStream.getResourceId());
+        // the input stream expired, close it
+        seekableInputStream.close();
+      }
+    }
+  }
+
+  /**
+   * Acquires an input stream. For seekable input streams, if there is an available input stream in
+   * the cache, reuse it and repositions the offset, otherwise the manager opens a new input stream.
+   *
+   * @param ufs the under file system
+   * @param path the path to the under storage file
+   * @param offset the offset to open
+   * @return the acquired input stream
+   * @throws IOException if the input stream fails to open
+   */
+  public InputStream acquire(UnderFileSystem ufs, String path, long offset) throws IOException {
+    if (!ufs.isSeekable() || !isCachingEnabled()) {
+      // not able to cache, always return a new input stream
+      return ufs.open(path, OpenOptions.defaults().setOffset(offset));
+    }
+
+    // explicit cache cleanup
+    mUnderFileInputStreamCache.cleanUp();
+
+    UfsInputStreamIds resources;
+    synchronized (mFileToInputStreamIds) {
+      if (mFileToInputStreamIds.containsKey(path)) {
+        resources = mFileToInputStreamIds.get(path);
+      } else {
+        resources = new UfsInputStreamIds();
+        mFileToInputStreamIds.put(path, resources);
+      }
+    }
+
+    synchronized (resources) {
+      long nextId = -1;
+      SeekableUnderFileInputStream inputStream = null;
+      // find the next available input stream from the cache
+      for (long id : resources.availableIds()) {
+        inputStream = mUnderFileInputStreamCache.getIfPresent(id);
+        if (inputStream != null) {
+          nextId = id;
+          LOG.debug("Reused the under file input stream resource of {}", nextId);
+          // for the cached ufs instream, seek to the requested position
+          inputStream.seek(offset);
+          break;
+        }
+      }
+      // no cached input stream is available, open a new one
+      if (nextId == -1) {
+        nextId = IdUtils.getRandomNonNegativeLong();
+        final long newId = nextId;
+        try {
+          inputStream = mUnderFileInputStreamCache.get(nextId, () -> {
+            SeekableUnderFileInputStream ufsStream = (SeekableUnderFileInputStream) ufs.open(path,
+                OpenOptions.defaults().setOffset(offset));
+            LOG.debug("Created the under file input stream resource of {}", newId);
+            ufsStream.setResourceId(newId);
+            ufsStream.setFilePath(path);
+            return ufsStream;
+          });
+        } catch (ExecutionException e) {
+          LOG.warn("Failed to retrieve the cached UFS instream");
+          // fall back to a ufs creation.
+          return ufs.open(path, OpenOptions.defaults().setOffset(offset));
+        }
+      }
+
+      // mark the input stream id as acquired
+      resources.acquire(nextId);
```

Comment:
We are adding a new value to `resources`, not acquiring an existing value? The API is a bit counterintuitive, since it looks like there is no way to populate the pool of ids.

Reply:
`nextId` can be an existing value, a new id is generated (line192-209) only when no existing resource is available.

Reply:
Looking at the API of `UfsInputStreamIds`, it is not obvious how to add a new id, since acquire implies that the id already existed. Although `acquire` doubles as adding a new id to the set, I find it counterintuitive, but its up to you.

Reply:
I see. I think it's fine because the `nextId` is created and added to the cache above. This line is for bookkeeping so by the time it's invoked the id already exists.